apps/assistants/models.py

```diff
@@ -171,29 +172,47 @@ def compare_with_model(self, new: Self, exclude_fields: list[str], early_abort=F
     def archive(self):
         from apps.assistants.tasks import delete_openai_assistant_task
 
-        # don't archive assistant if it's still referenced by an active experiment or pipeline
-        if self.get_related_experiments_queryset().exists():
+        if self.get_related_experiments_queryset().exists() or self.get_related_pipeline_node_queryset().exists():
             return False
-
-        if self.get_related_pipeline_node_queryset().exists():
-            return False
-
-        super().archive()
         if self.is_working_version:
-            # TODO: should this delete the assistant from OpenAI?
+            for (
+                version
+            ) in self.versions.all():  # first perform all checks so assistants are not archived prior to return False
+                if (
+                    version.get_related_experiments_queryset().exists()
+                    or version.get_related_pipeline_node_queryset().exists()
+                ):
+                    return False
+            for version in self.versions.all():
+                delete_openai_assistant_task.delay(version.id)
             self.versions.update(is_archived=True, audit_action=AuditAction.AUDIT)
-        else:
-            delete_openai_assistant_task.delay(self.id)
 
+        super().archive()
+        delete_openai_assistant_task.delay(self.id)
         return True
 
-    def get_related_experiments_queryset(self):
-        return self.experiment_set.filter(is_archived=False)
+    def get_related_experiments_queryset(self, assistant_ids: list = None):
+        if assistant_ids:
+            return Experiment.objects.filter(
+                Q(working_version_id=None) | Q(is_default_version=True),
+                assistant_id__in=assistant_ids,
+                is_archived=False,
+            )
+
+        return self.experiment_set.filter(Q(working_version_id=None) | Q(is_default_version=True), is_archived=False)
 
-    def get_related_pipeline_node_queryset(self):
+    def get_related_pipeline_node_queryset(self, assistant_ids: list = None):
```

Comment:
These should check the experiments that the pipeline belongs to instead of the pipeline itself?